File: src/lesson1/Sorts.java
package lesson1;

import java.util.Arrays;
import java.util.Calendar;
import java.util.Random;

@SuppressWarnings("WeakerAccess")
public class Sorts {

    public static <T extends Comparable<T>> void insertionSort(T[] elements) {
        for (int i = 1; i < elements.length; i++) {
            T current = elements[i];
            int j = i - 1;
            for (; j >= 0; j--) {
                if (elements[j].compareTo(current) > 0) elements[j+1] = elements[j];
                else break;
            }
            elements[j+1] = current;
        }
    }

    public static void insertionSort(int[] elements) {
        for (int i = 1; i < elements.length; i++) {
            int current = elements[i];
            int j = i - 1;
            for (; j >= 0; j--) {
                if (elements[j] > current) elements[j+1] = elements[j];
                else break;
            }
            elements[j+1] = current;
        }
    }

    private static void merge(int[] elements, int begin, int middle, int end) {
        int[] left = Arrays.copyOfRange(elements, begin, middle);
        int[] right = Arrays.copyOfRange(elements, middle, end);
        int li = 0, ri = 0;
        for (int i = begin; i < end; i++) {
            if (li < left.length && (ri == right.length || left[li] <= right[ri])) {
                elements[i] = left[li++];
            }
            else {
                elements[i] = right[ri++];
            }
        }
    }

    private static void mergeSort(int[] elements, int begin, int end) {
        if (end - begin <= 1) return;
        int middle = (begin + end) / 2;
        mergeSort(elements, begin, middle);
        mergeSort(elements, middle, end);
        merge(elements, begin, middle, end);
    }

    public static void mergeSort(int[] elements) {
        mergeSort(elements, 0, elements.length);
    }

    private static void heapify(int[] elements, int start, int length) {
        int left = 2 * start + 1;
        int right = left + 1;
        int max = start;
        if (left < length && elements[left] > elements[max]) {
            max = left;
        }
        if (right < length && elements[right] > elements[max]) {
            max = right;
        }
        if (max != start) {
            int temp = elements[max];
            elements[max] = elements[start];
            elements[start] = temp;
            heapify(elements, max, length);
        }
    }

    private static void buildHeap(int[] elements) {
        for (int start = elements.length / 2 - 1; start >= 0; start--) {
            heapify(elements, start, elements.length);
        }
    }

    public static void heapSort(int[] elements) {
        buildHeap(elements);
        for (int j = elements.length - 1; j >= 1; j--) {
            int temp = elements[0];
            elements[0] = elements[j];
            elements[j] = temp;
            heapify(elements, 0, j);
        }
    }

    private static final Random random = new Random(Calendar.getInstance().getTimeInMillis());

    private static int partition(int[] elements, int min, int max) {
        int x = elements[min + random.nextInt(max - min + 1)];
        int left = min, right = max;
        while (left <= right) {
            while (elements[left] < x) {
                left++;
            }
            while (elements[right] > x) {
                right--;
            }
            if (left <= right) {
                int temp = elements[left];
                elements[left] = elements[right];
                elements[right] = temp;
                left++;
                right--;
            }
        }
        return right;
    }

    private static void quickSort(int[] elements, int min, int max) {
        if (min < max) {
            int border = partition(elements, min, max);
            quickSort(elements, min, border);
            quickSort(elements, border + 1, max);
        }
    }

    public static void quickSort(int[] elements) {
        quickSort(elements, 0, elements.length - 1);
    }

    public static int[] countingSort(int[] elements, int limit) {
        int[] count = new int[limit + 1];
        for (int element: elements) {
            count[element]++;
        }
        for (int j = 1; j <= limit; j++) {
            count[j] += count[j - 1];
        }
        int[] out = new int[elements.length];
        for (int j = elements.length - 1; j >= 0; j--) {
            out[count[elements[j]] - 1] = elements[j];
            count[elements[j]]--;
        }
        return out;
    }
}

File: src/lesson1/Sorts.kt
package lesson1

import java.util.*

private val random = Random(Calendar.getInstance().timeInMillis)

fun <T : Comparable<T>> insertionSort(elements: MutableList<T>) {
    for (i in 1 until elements.size) {
        val current = elements[i]
        var j = i - 1
        while (j >= 0) {
            if (elements[j] > current)
                elements[j + 1] = elements[j]
            else
                break
            j--
        }
        elements[j + 1] = current
    }
}

private fun merge(elements: IntArray, begin: Int, middle: Int, end: Int) {
    val left = elements.copyOfRange(begin, middle)
    val right = elements.copyOfRange(middle, end)
    var li = 0
    var ri = 0
    for (i in begin until end) {
        if (li < left.size && (ri == right.size || left[li] <= right[ri])) {
            elements[i] = left[li++]
        } else {
            elements[i] = right[ri++]
        }
    }
}

private fun mergeSort(elements: IntArray, begin: Int, end: Int) {
    if (end - begin <= 1) return
    val middle = (begin + end) / 2
    mergeSort(elements, begin, middle)
    mergeSort(elements, middle, end)
    merge(elements, begin, middle, end)
}

fun mergeSort(elements: IntArray) {
    mergeSort(elements, 0, elements.size)
}

private fun heapify(elements: IntArray, start: Int, length: Int) {
    val left = 2 * start + 1
    val right = left + 1
    var max = start
    if (left < length && elements[left] > elements[max]) {
        max = left
    }
    if (right < length && elements[right] > elements[max]) {
        max = right
    }
    if (max != start) {
        val temp = elements[max]
        elements[max] = elements[start]
        elements[start] = temp
        heapify(elements, max, length)
    }
}

private fun buildHeap(elements: IntArray) {
    for (start in elements.size / 2 - 1 downTo 0) {
        heapify(elements, start, elements.size)
    }
}

fun heapSort(elements: IntArray) {
    buildHeap(elements)
    for (j in elements.size - 1 downTo 1) {
        val temp = elements[0]
        elements[0] = elements[j]
        elements[j] = temp
        heapify(elements, 0, j)
    }
}

private fun partition(elements: IntArray, min: Int, max: Int): Int {
    val x = elements[min + random.nextInt(max - min + 1)]
    var left = min
    var right = max
    while (left <= right) {
        while (elements[left] < x) {
            left++
        }
        while (elements[right] > x) {
            right--
        }
        if (left <= right) {
            val temp = elements[left]
            elements[left] = elements[right]
            elements[right] = temp
            left++
            right--
        }
    }
    return right
}

private fun quickSort(elements: IntArray, min: Int, max: Int) {
    if (min < max) {
        val border = partition(elements, min, max)
        quickSort(elements, min, border)
        quickSort(elements, border + 1, max)
    }
}

fun quickSort(elements: IntArray) {
    quickSort(elements, 0, elements.size - 1)
}

fun countingSort(elements: IntArray, limit: Int): IntArray {
    val count = IntArray(limit + 1)
    for (element in elements) {
        count[element]++
    }
    for (j in 1..limit) {
        count[j] += count[j - 1]
    }
    val out = IntArray(elements.size)
    for (j in elements.indices.reversed()) {
        out[count[elements[j]] - 1] = elements[j]
        count[elements[j]]--
    }
    return out
}

/**
 * Библиотечные сортировки (приведены только для примера,
 * рекомендуется использовать list.sorted() и list.sort() напрямую)
 */
fun <T : Comparable<T>> librarySortForImmutable(list: List<T>): List<T> = list.sorted()

fun <T : Comparable<T>> librarySortForMutable(list: MutableList<T>) {
    list.sort()
}